### PR TITLE
Fixes #12168: Port sudoers configuration technique to multiversionned technique

### PR DESCRIPTION
**Pull request by ncharles**
https://www.rudder-project.org/redmine/issues/12168

### CODE REVIEW

**Comment by ncharles** [comment]
Commit modified

**Comment by ncharles** [comment]
Commit modified

File: techniques/system/common/1.0/hooks.st [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+      "na_report"
+        usebundle  => _rudder_common_report_hooks("${reporting[${reportkeys}][technique]}", "result_na", "${reporting[${reportkeys}][id]}", "${reporting[${reportkeys}][name]}", "${reporting[${reportkeys}][value]}", "Installing package ${parameters[name]}", "${reporting[${reportkeys}][mode]}"),
+        ifvarclass => "!${class_condition}";
+
+      "clean_reporting_context_${reporting[0][id]}"
+        usebundle  => clean_reporting_context;
+
+  reports:
+    inform.pass2::
```
Comment (amousset) [inline techniques/system/common/1.0/hooks.st]: :+1: we need more of these in Rudder

File: techniques/system/common/1.0/hooks.st [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+      "set_dry_mode"  string => "false";
+
+    pass1.!is_enforce::
+      "set_dry_mode"  string => "true";
+
+  classes:
+      # define if there is at least one enforce
+      "is_enforce"            expression => strcmp("${reporting[${reportkeys}][mode]}", "enforce");
+
+       # For reporting, detect each enforce/audit
```
Comment (amousset) [inline techniques/system/common/1.0/hooks.st]: identation is 1 char off here

File: techniques/system/common/1.0/hooks.st (hunk 1 above)
Comment (ncharles) [reply techniques/system/common/1.0/hooks.st]: Yes !

File: techniques/system/common/1.0/hooks.st [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+      # For reporting, detect each enforce/audit
+      "is_enforce_${reportkeys}" expression => strcmp("${reporting[${reportkeys}][mode]}", "enforce");
+
+    is_enforce::
+      # Check the version of visudo; if before 1.7 (so 1.6 or less), we need to use strict mode
+      # For version 1.7, bugs in the strict parsing prevent from using safely the strict mode (see http://www.sudo.ws/bugs/show_bug.cgi?id=519 )
+      # Versions after 1.7 don't need to have Aliases defined before they are used, and strict mode only checks that, so we can safely ignore it
+      # visudo -V can returns several lines, so we need to keep only the visudo version line
+      "visudo_use_strict_mode" expression => returnszero("/usr/sbin/visudo  -V | grep 'visudo version' | /bin/sed -e 's/visudo version \([0-9]\+\)\.\([0-9]\+\)\..*/\1 \2/p' | /usr/bin/awk '{ if(($1<1)||($1==1 \&\& $2<=6)) { exit 0 }; exit 1 }'", "useshell");
```
Comment (amousset) [inline techniques/system/common/1.0/hooks.st]: :open_mouth:

File: techniques/systemSettings/userManagement/sudoParameters/3.2/sudoParameters.st [hunk 4]
```diff
@@ -17,12 +17,14 @@
 #####################################################################################
 
 ##########################################################################
-# Sudo configuration Technique                                                  #
+# Sudo configuration Technique                                           #
 # ---------------------------------------------------------------------- #
 # Objective : Configure /etc/sudoers according to the given parameters   #
 ##########################################################################
```
Comment (amousset) [inline techniques/systemSettings/userManagement/sudoParameters/3.2/sudoParameters.st]: Could you add a comment to explain the general behavior (tmp file vs? audit mode)?

**Comment by ncharles** [comment]
Commit modified

**Pull request merged**